test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java

```diff
@@ -0,0 +1,169 @@
+import static javax.lang.model.SourceVersion.*;
+import javax.lang.model.element.*;
+import javax.lang.model.util.*;
+
+/**
+ * Test Elements.{isCompactConstructor, isCanonicalConstructor}.
+ */
+public class TestRecordPredicates extends JavacTestingAbstractProcessor {
+    public boolean process(Set<? extends TypeElement> annotations,
+                           RoundEnvironment roundEnv) {
+        if (!roundEnv.processingOver()) {
+            // Test null handling
+            Elements vacuousElements = new VacuousElements();
+            expectFalse( () -> vacuousElements.isCompactConstructor(null));
+            expectFalse( () -> vacuousElements.isCanonicalConstructor(null));
+
+            expectNpe( () -> elements.isCompactConstructor(null));
+            expectNpe( () -> elements.isCanonicalConstructor(null));
+
+            for (var typeElt :
+                     ElementFilter.typesIn(roundEnv.getElementsAnnotatedWith(ExpectedPredicates.class))) {
+                ExpectedPredicates ep = typeElt.getAnnotation(ExpectedPredicates.class);
+                for (ExecutableElement ctor : ElementFilter.constructorsIn(typeElt.getEnclosedElements())) {
+                    boolean isCompact = elements.isCompactConstructor(ctor);
+                    if (isCompact != ep.isCompact()) {
+                        messager.printError("Unexpected isCompact value on ", ctor);
+                    }
+
+                    boolean isCanonical = elements.isCanonicalConstructor(ctor);
+                    if (isCanonical != ep.isCanonical()) {
+                        messager.printError("Unexpected isCanonical value on ", ctor);
+                    }
+
+                    if (isCompact && !isCanonical) {
+                        messager.printError("Compact constructors not reported as canonical ", ctor);
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    private void expectNpe(java.util.function.BooleanSupplier bs) {
+        try {
+            bs.getAsBoolean();
+            messager.printError("Did not get expected NPE");
+        } catch (NullPointerException npe) {
+                ; // Expected
+        }
+    }
+
+    private void expectFalse(java.util.function.BooleanSupplier bs) {
+        try {
+            boolean result = bs.getAsBoolean();
+            if (result) {
+                messager.printError("Unexpected true result");
+            }
+        } catch (NullPointerException npe) {
+            messager.printError("Unexpected NPE thrown");
+        }
+    }
+
+
+    /**
+     * Annotation the class, not just constructor, since many of the
+     * constructors that are of interest are implicitly declared.
+     */
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface ExpectedPredicates {
+        boolean isCompact()   default false;
+        boolean isCanonical() default false;
+    }
+
+    @ExpectedPredicates(isCompact=true, isCanonical=true)
```

Comment:
side: this is uncovering a bug in the compiler, isCanonical should be false in this case. I have filed JDK-8291914 to address this issue

Reply:
Part of the regression tests for  JDK-8291914 already written then :-)

Reply:
:)